### ISSUE TITLE
maximum and nan / identity for maximum (migrated from Trac #241)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#550]
Original ticket http://projects.scipy.org/numpy/ticket/241
Reported 2006-08-11 by trac user martin_wiechert, assigned to unknown.

How should maximum handle nans? I think there are only two reasonable behaviours. Either nans should be ignored or the result should be nan.
(I personally favour a keyword switch over a separate function nanmax.)

In the current implementation the outcome depends on the ''order'' of elements:
    >>> import numpy
    >>> numpy.version.version
    '1.0b1'
    >>> from numpy import *
    >>> a = array ((1, nan))
    >>> maximum.reduce (a)
    1.0
    >>> maximum.reduce (a [::-1])
    nan
This is certainly confusing if not a bug.

As a related request I'd like to see an identity element for maximum, namely -inf.

All the above of course also holds for minimum.

**Comment by thouis** [comment]
Comment in Trac by trac user sasha, 2006-08-11

Note that the observed inconsistency is not specific to numpy:

```
Python 2.4.2 (#2, Jan 13 2006, 12:00:38)
[GCC 3.4.4 20050721 (Red Hat 3.4.4-2)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> nan = float('nan')
>>> max(nan,1)
nan
>>> max(1,nan)
1
```

Numpy should stop pretending that it supports nans despite the fact that python does not.

**Comment by thouis** [comment]
Comment in Trac by @XXXthouis not really:teoliphant, 2006-08-13

NaN's are a peculiar creature.  Their behavior on comparison can be confusing but it is consistent.  Numpy just inherits the way the platform treats nans.  It doesn't try to do anything special with them by default as that would slow down all code (even code where the user knows nans are not going to show up).  

Handling this correctly is left to the user at this point.

**Comment by thouis** [comment]
Comment in Trac by trac user peridot, 2006-10-24

Replying to [comment:2 oliphant]:

> NaN's are a peculiar creature.  Their behavior on comparison can be confusing but it is consistent.  Numpy just inherits the way the platform treats nans.  It doesn't try to do anything special with them by default as that would slow down all code (even code where the user knows nans are not going to show up).  

Unfortunately, the right behaviour for maximum would be to return a nan if any element is a nan. The current behaviour comes from the fact that nan returns false for all comparisons, so in fact what happens now is that the nan is ignored unless it's in the first place, in which case the result is nan. It would be easy to check for this.
    In [9]: minimum.reduce(array([1,2,3,nan,2]))
    Out[9]: 1.0

```
In [10]: minimum.reduce(array([nan,1,2,3,2]))
Out[10]: nan

In [11]: minimum.reduce(array([inf,nan,1,2,3,2]))
Out[11]: 1.0
```

In fact, a fix for this would be to use -inf as the initial (identity) value; another would be to set nans to raise an exception, which could then be trapped and handled specially before returning the exception state to normal.

It'd be nice to avoid secret disappearing nans, but if that's infeasible, I think removing the special behaviour when nan is first would help.

**Comment by thouis** [comment]
Comment in Trac by @XXXthouis not really:alberts, 2007-05-12

**Comment by thouis** [comment]
Comment in Trac by @XXXthouis not really:cournape, 2009-02-19

Fixed in truc (upcoming 1.3).
